File: python/exercises/practice/react/react.py
class InputCell:
    def __init__(self, initial_value):
        self._value = initial_value
        self.consumers = []

    @property
    def value(self):
        return self._value

    @value.setter
    def value(self, new_value):
        if self._value != new_value:
            self._value = new_value
            propagate(self)


class ComputeCell:
    def __init__(self, inputs, compute_function):
        self.inputs = inputs
        self.compute_function = compute_function
        self.consumers = []
        self.callbacks = []

        for input_cell in self.inputs:
            input_cell.consumers.append(self)

        self._value = self.compute_function([inp.value for inp in self.inputs])

    @property
    def value(self):
        return self._value

    def add_callback(self, callback):
        self.callbacks.append(callback)

    def remove_callback(self, callback):
        if callback in self.callbacks:
            self.callbacks.remove(callback)


def get_reachable_compute_cells(start_cell):
    reachable = set()
    queue = [start_cell]
    while queue:
        curr = queue.pop(0)
        for consumer in curr.consumers:
            if consumer not in reachable:
                reachable.add(consumer)
                queue.append(consumer)
    return reachable


def topological_sort(reachable_cells):
    in_degree = {cell: 0 for cell in reachable_cells}
    for cell in reachable_cells:
        for consumer in cell.consumers:
            if consumer in reachable_cells:
                in_degree[consumer] += 1

    queue = [cell for cell, deg in in_degree.items() if deg == 0]
    topo_order = []

    while queue:
        curr = queue.pop(0)
        topo_order.append(curr)
        for consumer in curr.consumers:
            if consumer in reachable_cells:
                in_degree[consumer] -= 1
                if in_degree[consumer] == 0:
                    queue.append(consumer)

    return topo_order


def propagate(start_cell):
    reachable = get_reachable_compute_cells(start_cell)
    if not reachable:
        return

    topo_order = topological_sort(reachable)

    # Record initial values
    initial_values = {cell: cell.value for cell in topo_order}

    # Update values in topological order
    for cell in topo_order:
        cell._value = cell.compute_function([inp.value for inp in cell.inputs])

    # Trigger callbacks for cells whose values changed
    for cell in topo_order:
        if cell.value != initial_values[cell]:
            for callback in list(cell.callbacks):
                callback(cell.value)

File: java/exercises/practice/react/src/main/java/React.java
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.IdentityHashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Queue;
import java.util.Set;
import java.util.function.Consumer;
import java.util.function.Function;

public class React {

    public static class Cell<T> {
        protected T value;
        final List<ComputeCell<?>> consumers = new ArrayList<>();
        int level = 0;

        public T getValue() {
            return value;
        }

        int getLevel() {
            return level;
        }
    }

    public static class InputCell<T> extends Cell<T> {
        public InputCell(T initialValue) {
            this.value = initialValue;
            this.level = 0;
        }

        public void setValue(T newValue) {
            if (Objects.equals(this.value, newValue)) {
                return;
            }
            this.value = newValue;
            propagate();
        }

        private void propagate() {
            // Collect all reachable compute cells downstream
            Set<ComputeCell<?>> reachable = new LinkedHashSet<>();
            Queue<Cell<?>> queue = new ArrayDeque<>();
            queue.add(this);

            while (!queue.isEmpty()) {
                Cell<?> current = queue.poll();
                for (ComputeCell<?> consumer : current.consumers) {
                    if (reachable.add(consumer)) {
                        queue.add(consumer);
                    }
                }
            }

            if (reachable.isEmpty()) {
                return;
            }

            // Capture stable state values prior to this update cycle
            Map<ComputeCell<?>, Object> oldValues = new IdentityHashMap<>();
            for (ComputeCell<?> cell : reachable) {
                oldValues.put(cell, cell.getValue());
            }

            // Topological sort by DAG depth level to guarantee dependencies resolve first
            List<ComputeCell<?>> sortedCells = new ArrayList<>(reachable);
            sortedCells.sort(Comparator.comparingInt(Cell::getLevel));

            // Recompute values in topological order to bring all cells into the new stable state
            for (ComputeCell<?> cell : sortedCells) {
                cell.recompute();
            }

            // Fire callbacks only for cells whose value changed between stable states
            for (ComputeCell<?> cell : sortedCells) {
                Object oldValue = oldValues.get(cell);
                Object newValue = cell.getValue();
                if (!Objects.equals(oldValue, newValue)) {
                    cell.fireCallbacks();
                }
            }
        }
    }

    public static class ComputeCell<T> extends Cell<T> {
        private final Function<List<T>, T> function;
        private final List<Cell<T>> cells;
        private final List<Consumer<T>> callbacks = new ArrayList<>();

        public ComputeCell(Function<List<T>, T> function, List<Cell<T>> cells) {
            this.function = function;
            this.cells = new ArrayList<>(cells);

            int maxLevel = 0;
            Set<Cell<T>> uniqueCells = new LinkedHashSet<>(cells);
            for (Cell<T> cell : uniqueCells) {
                maxLevel = Math.max(maxLevel, cell.getLevel());
                cell.consumers.add(this);
            }
            this.level = maxLevel + 1;
            this.value = computeValue();
        }

        void recompute() {
            this.value = computeValue();
        }

        private T computeValue() {
            List<T> values = new ArrayList<>(cells.size());
            for (Cell<T> cell : cells) {
                values.add(cell.getValue());
            }
            return function.apply(values);
        }

        public void addCallback(Consumer<T> callback) {
            callbacks.add(callback);
        }

        public void removeCallback(Consumer<T> callback) {
            callbacks.remove(callback);
        }

        void fireCallbacks() {
            List<Consumer<T>> snapshot = new ArrayList<>(callbacks);
            for (Consumer<T> callback : snapshot) {
                if (callbacks.contains(callback)) {
                    callback.accept(this.value);
                }
            }
        }
    }

    public static <T> InputCell<T> inputCell(T initialValue) {
        return new InputCell<>(initialValue);
    }

    public static <T> ComputeCell<T> computeCell(Function<List<T>, T> function, List<Cell<T>> cells) {
        return new ComputeCell<>(function, cells);
    }
}
